File: src/kingdom/xp.ts
export function calculateHexXP(
    {
        hexes,
        kingdomSize,
        useVK,
        xpPerClaimedHex,
    }: {
        hexes: number,
        xpPerClaimedHex: number,
        kingdomSize: number,
        useVK: boolean
    },
): number {
    if (useVK) {
        if (kingdomSize < 10) {
            return hexes * 100;
        } else if (kingdomSize < 25) {
            return hexes * 50;
        } else if (kingdomSize < 50) {
            return hexes * 25;
        } else if (kingdomSize < 100) {
            return hexes * 10;
        } else {
            return hexes * 5;
        }
    } else {
        return hexes * xpPerClaimedHex;
    }
}

export function calculateRpXP(
    {
        rp,
        kingdomLevel,
        rpToXpConversionRate,
        rpToXpConversionLimit,
        useVK,
    }: {
        rp: number,
        kingdomLevel: number,
        rpToXpConversionRate: number,
        rpToXpConversionLimit: number,
        useVK: boolean
    },
): number {
    let xp = 0;
    if (useVK) {
        if (kingdomLevel < 5) {
            xp = rp * 10;
        } else if (kingdomLevel < 9) {
            xp = rp * 7;
        } else if (kingdomLevel < 13) {
            xp = rp * 5;
        } else if (kingdomLevel < 17) {
            xp = rp * 2;
        }
    } else {
        xp = rp * rpToXpConversionRate;
    }
    return Math.min(rpToXpConversionLimit, xp);
}

export function calculateEventXP(modifier: number): number {
    const xp: Record<string, number> = {
        '-4': 10,
        '-3': 15,
        '-2': 20,
        '-1': 30,
        '0': 40,
        '1': 60,
        '2': 80,
        '3': 120,
        '4': 160,
    };
    const key = `${modifier}`;
    if (key in xp) {
        return xp[key];
    } else {
        return 0;
    }
}

File: src/commonMain/kotlin/at/posselt/pfrpg2e/data/kingdom/Xp.kt
package at.posselt.pfrpg2e.data.kingdom

import kotlin.math.min

fun calculateHexXP(
    hexes: Int,
    xpPerClaimedHex: Int,
    kingdomSize: Int,
    useVK: Boolean
) = if (useVK) {
    when {
        kingdomSize < 10 -> hexes * 100
        kingdomSize < 25 -> hexes * 50
        kingdomSize < 50 -> hexes * 25
        kingdomSize < 100 -> hexes * 10
        else -> hexes * 5
    }
} else {
    hexes * xpPerClaimedHex;
}

fun calculateRpXP(
    rp: Int,
    kingdomLevel: Int,
    rpToXpConversionRate: Int,
    rpToXpConversionLimit: Int,
    useVK: Boolean,
): Int {
    val xp = if (useVK) {
        when {
            kingdomLevel < 5 -> rp * 10
            kingdomLevel < 9 -> rp * 7
            kingdomLevel < 13 -> rp * 5
            kingdomLevel < 17 -> rp * 2
            else -> rp
        }
    } else {
        rp * rpToXpConversionRate
    }
    return min(rpToXpConversionLimit, xp)
}

fun calculateEventXP(modifier: Int) =
    when (modifier) {
        -4 -> 10
        -3 -> 15
        -2 -> 20
        -1 -> 30
        0 -> 40
        1 -> 60
        2 -> 80
        3 -> 120
        4 -> 160
        else -> 0
    }
